databricks/koalas/tests/test_dataframe.py

```diff
@@ -3895,3 +3895,71 @@ def test_first_valid_index(self):
         pdf = pd.Series([]).to_frame()
         kdf = ks.Series([]).to_frame()
         self.assert_eq(pdf.first_valid_index(), kdf.first_valid_index())
+
+    def test_product(self):
+        pdf = pd.DataFrame(
+            {"A": [1, 2, 3, 4, 5], "B": [10, 20, 30, 40, 50], "C": ["a", "b", "c", "d", "e"]}
+        )
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index())
+
+        # Named columns
+        pdf.columns.name = "Koalas"
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index())
+
+        # MultiIndex columns
+        pdf.columns = pd.MultiIndex.from_tuples([("a", "x"), ("b", "y"), ("c", "z")])
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index())
+
+        # Named MultiIndex columns
+        pdf.columns.names = ["Hello", "Koalas"]
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index())
+
+        # No numeric columns
+        pdf = pd.DataFrame({"key": ["a", "b", "c"], "val": ["x", "y", "z"]})
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index())
+
+        # No numeric named columns
+        pdf.columns.name = "Koalas"
+        kdf = ks.from_pandas(pdf)
+        self.assert_eq(pdf.prod(), kdf.prod().sort_index(), almost=True)
```

Comment:
Revert this change since couldn't reproduce the failure below in my local environment.

<img width="672" alt="스크린샷 2020-09-08 오후 12 36 26" src="https://user-images.githubusercontent.com/44108233/92430490-f0480f00-f1cf-11ea-9b5a-6f848ccf9425.png">

Actually, even the name of Index in the result looks not different for me.

```python
>>> pdf
Koalas key val
0        a   x
1        b   y
2        c   z

>>> pdf.prod()
Series([], dtype: float64)
>>> kdf.prod()
Series([], dtype: float64)

>>> pdf.prod().index.name  # None
>>> kdf.prod().index.name  # None
```

Reply:
It IS different in my local:

```py
>>> pdf = pd.DataFrame({"key": ["a", "b", "c"], "val": ["x", "y", "z"]})
>>> pdf.columns.name = "Koalas"
>>> pdf
Koalas key val
0        a   x
1        b   y
2        c   z
>>> pdf.prod()
Series([], dtype: float64)
>>> pdf.prod().index
Index([], dtype='object', name='Koalas')
>>> pdf.prod().index.names
FrozenList(['Koalas'])
```

Reply:
Thanks for the sharing, @ueshin !

Oh this behavior seems to be changed from pandas 1.1.0.

Then I think maybe we can leave this as it is because we should match our behavior to the latest pandas ??

Reply:
ah, I see. sounds like inconsistent behavior, though..

```py
>>> pdf = pd.DataFrame({"key": [1, 2, 3], "val": [4, 5, 6]})
>>> pdf.columns.name = "Koalas"
>>> pdf.prod().index
Index(['key', 'val'], dtype='object', name='Koalas')

>>> pdf = pd.DataFrame({"key": ["a", "b", "c"], "val": ["x", "y", "z"]})
>>> pdf.columns.name = "Koalas"
>>> pdf.prod().index
Index([], dtype='object')
```

Reply:
Ah yeah, they only preserve the name of index when the result is not empty Series.